src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs

```diff
@@ -255,13 +269,25 @@ private async Task<IEnumerable<PackageReference>> GetPackageReferencesAsync(Pack
                 }
                 _lastTimeAssetsModified = fileInfo.LastWriteTimeUtc;
                 _lastPackageSpec = new WeakReference<PackageSpec>(packageSpec);
+
+                // clear the transitive packages cache, since we don't know when a dependency has been removed
+                _transitivePackages = new List<(NuGetFramework, Dictionary<string, ProjectInstalledPackage>)>();
```

Comment:
There is no other mechanism to remove packages from the cache when then are uninstalled. Instead of keeping track of what has changed, I decided it would be better to rebuild it from scratch when the assets file has changed. 
This is different from the way the installed packages work, since we can get the current list from the packageSpec that is passed in. I couldn't find the transitive packages in the packageSpec, so I can't use that to determine what transitive packages are currently present. If they are in there, they are well hidden :). 
One thing to note here is that packages don't get removed from the installed packages cache when they are uninstalled. They are just ignored if they are not present in the packageSpec.

Reply:
I don't see anything getting removed from _installedPackages. When I uninstall packages, the aren[t removed from _installedPackages. The correct direct dependencies are returned since _installedPackages is only used to look up the installed version of the packages in the package spec.
For transitive packages, I didn't find them in the package spec, so I rely on _transitivePackages to get the list of packages and versions.

Reply:
if you are seeing that, then we likely have a bug :) 

The following
`GetInstalledPackages`
Uninstall package
`GetInstalledPackages` should return different results.

Reply:
GetInstalledPackagesAsync returns the correct results after a package is uninstalled. However, _installedPackages still contains the uninstalled package. This works because _installedPackages is only used to look up the installed version, not to determine which packages are installed. 
So, while the results are correct, does still keep info packages that were uninstalled.

Reply:
I see. 
All good then. As long as we're returning the correct thing.